### PR TITLE
Deprecate the `PROJECT_ROOT_FOLDER` and `PROJECT_NAME` environment variables

### LINKED ISSUE #205: Remove the need for PROJECT_ROOT_DIR env var
We should stop relying on this env var and requiring it to be set as we could instead be able to detect that project root automatically in the actions that needs it.

At first I thought we could make use of something like `git rev-parse --show-toplevel` to detect that project root path automatically. But then I realized that I think that won't even be necessary, because [as documented in the _fastlane_ docs](https://docs.fastlane.tools/advanced/fastlane/#directory-behavior), the working directory is already guaranteed to be the project root itself for code run from actions defined in plugin like the `release-toolkit`. So we can probably just replace it with `.` and use relative paths, as the `Dir.pwd` should already be the project root folder in our context!

### PR DESCRIPTION
**Pull request by spencertransier**
## What does it do?

Fixes https://github.com/wordpress-mobile/release-toolkit/issues/205

This PR removes the reliance on the `PROJECT_ROOT_FOLDER` and `PROJECT_NAME` environment variables across the Release Toolkit actions and helpers. 

That env var is primarily used for pointing to the `build.gradle` or `version.properties` files for Android versioning. I added deprecation warnings to some shared ConfigItems are added to each of the actions that rely on one or both of the env vars. 

## Checklist before requesting a review

- [X] Run `bundle exec rubocop` to test for code style violations and recommendations
- [x] Add Unit Tests (aka `specs/*_spec.rb`) if applicable
- [X] Run `bundle exec rspec` to run the whole test suite and ensure all your tests pass
- [x] Make sure you added an entry in [the `CHANGELOG.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/CHANGELOG.md#trunk) to describe your changes under the appropriate existing `###` subsection of the existing `## Trunk` section.
- [x] If applicable, add an entry in [the `MIGRATION.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/MIGRATION.md) to describe how the changes will affect the migration from the previous major version and what the clients will need to change and consider.

### CODE REVIEW

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_betabuild_prechecks.rb [hunk 1]
```diff
@@ -8,18 +8,30 @@ def self.run(params)
...
+        build_gradle_path = params[:build_gradle_path]
+        version_properties_path = params[:version_properties_path]
+
         # Checkout default branch and update
         default_branch = params[:default_branch]
         Fastlane::Helper::GitHelper.checkout_and_pull(default_branch)
 
         # Check versions
-        release_version = Fastlane::Helper::Android::VersionHelper.get_release_version
+        release_version = Fastlane::Helper::Android::VersionHelper.get_release_version(
```
Comment (jkmassel) [inline lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_betabuild_prechecks.rb:21]: Can we use named parameters here? Makes it much easier to do backcompat (for instance, if we want to re-order the parameters), and easier to follow along in general.

Same for basically every instance of passing these objects around :)

**Review by jkmassel — CHANGES_REQUESTED** [review]
This is an excellent start, though I think we're adding a lot of nullability (which can bite us later).

IMHO, it'd be preferable to repeat ourselves a little here to remove it – for instance, we could add a (deprecated) `FastlaneCore::ConfigItem` that uses `ENV['PROJECT_ROOT_FOLDER']` as its default value to the relevant actions (bonus points: IINM we can specify that some arguments are incompatible with others – if someone uses `build_gradle_path` _and_ the environment-var-backed one, we can throw up an error).

This would allow anything underneath to always assume a value – for the most part we treat `Helpers` as private API, but not having to go back and adjust their signatures later is IMHO preferable. 

If we need to stick with the current "add a bunch of nullable arguments" for good reasons, I'd like to see tests to cover the different paths through those cases – if there's no nullability (thus a simpler code path all around) the implementation should change very little so hopefully the existing tests could cover it?

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_betabuild_prechecks.rb (hunk 1 above)
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_betabuild_prechecks.rb:21]: Added named parameters here: https://github.com/wordpress-mobile/release-toolkit/commit/73b6537b2dd20f5fb7f8bf6f3c126febb5d48cfb

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_build_prechecks.rb [hunk 2]
```diff
@@ -9,9 +9,14 @@ def self.run(params)
...
 
+        project_root_folder = params[:project_root_folder]
+        project_name = params[:project_name]
+        build_gradle_path = params[:build_gradle_path] || File.join(project_root_folder || '.', project_name, 'build.gradle')
+        version_properties_path = params[:version_properties_path] || File.join(project_root_folder || '.', 'version.properties')
+
         message = ''
-        beta_version = Fastlane::Helper::Android::VersionHelper.get_release_version unless !params[:beta] && !params[:final]
-        alpha_version = Fastlane::Helper::Android::VersionHelper.get_alpha_version if params[:alpha]
+        beta_version = Fastlane::Helper::Android::VersionHelper.get_release_version(build_gradle_path: build_gradle_path, version_properties_path: version_properties_path) unless !params[:beta] && !params[:final]
```
Comment (jkmassel) [inline lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_build_prechecks.rb]: Not sure how Rubocop feels about it, but should these params each have their own line?

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/deprecated.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  module Helper
+    # A helper class to store deprecated methods and actions
+    class Deprecated
+      # Creates a project_root_folder Fastlane ConfigItem
+      #
+      # @return [FastlaneCore::ConfigItem] The Fastlane ConfigItem for the `PROJECT_ROOT_FOLDER` environment variable
+      #
+      def self.project_root_folder_config_item
+        verify_block = proc do
+          unless ENV['PROJECT_ROOT_FOLDER'].nil?
```
Comment (jkmassel) [inline lib/fastlane/plugin/wpmreleasetoolkit/helper/deprecated.rb]: IINM, doesn't this block only run if `params[:project_root_folder]` is requested?

ie – do we need the `unless` wrapper?

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/deprecated.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+        FastlaneCore::ConfigItem.new(
+          key: :project_root_folder,
+          env_name: 'PROJECT_ROOT_FOLDER',
+          description: 'The path to the project root folder',
+          deprecated: true,
+          optional: true,
+          verify_block: verify_block,
+          type: String,
+          default_value: ENV['PROJECT_ROOT_FOLDER']
```
Comment (jkmassel) [inline lib/fastlane/plugin/wpmreleasetoolkit/helper/deprecated.rb]: Given the fact that this is picked up from `env_name`, I'm not sure we want it to be the default value like this?

WDYT?

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_codefreeze_prechecks.rb [hunk 5]
```diff
@@ -64,6 +78,18 @@ def self.available_options
                                        description: 'Default branch of the repository',
                                        type: String,
                                        default_value: Fastlane::Helper::GitHelper::DEFAULT_GIT_BRANCH),
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:version_properties_path]),
```
Comment (jkmassel) [inline lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_codefreeze_prechecks.rb]: IINM, this also conflicts with a developer using `project_root_folder`, no?

**Review by jkmassel — CHANGES_REQUESTED** [review]
Getting closer – assume all comments apply to each instance of the pattern (ie – if I commented on one config item, it should apply to all config items that share the name)

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_codefreeze_prechecks.rb (hunk 5 above)
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_codefreeze_prechecks.rb]: Correct, along with `project_name` as well. Addressed in https://github.com/wordpress-mobile/release-toolkit/commit/4af6d4e2cb1be043e27ddcf32fcc6c145a6fd441

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/deprecated.rb (hunk 4 above)
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/deprecated.rb]: Good point, we really don't need a default value for the `project_root_folder` or `project_name` config items. Addressed in https://github.com/wordpress-mobile/release-toolkit/commit/7d0240c2b7fcaf6a276a139e1b6a7a9bac6af6e2

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/deprecated.rb (hunk 3 above)
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/deprecated.rb]: Good point, fixed in https://github.com/wordpress-mobile/release-toolkit/commit/bd6b643d381f1fd5998af86b677a147d6975f63a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_build_prechecks.rb (hunk 2 above)
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_build_prechecks.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/commit/0277441e789efbb230d6b6d971b1d07e1548ae1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_tag_build.rb [hunk 6]
```diff
@@ -30,6 +41,18 @@ def self.available_options
...
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
+          FastlaneCore::ConfigItem.new(key: :version_properties_path,
+                                       description: 'Path to the version.properties file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:build_gradle_path, :project_name, :project_root_folder]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_tag_build.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_tag_build.rb [hunk 7]
```diff
@@ -30,6 +41,18 @@ def self.available_options
                                        description: 'True to skip tagging the alpha version',
                                        type: Boolean,
                                        default_value: true),
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_tag_build.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_release_version.rb [hunk 8]
```diff
@@ -19,7 +28,20 @@ def self.details
...
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
+          FastlaneCore::ConfigItem.new(key: :version_properties_path,
+                                       description: 'Path to the version.properties file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:build_gradle_path, :project_name, :project_root_folder]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_release_version.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_alpha_version.rb [hunk 9]
```diff
@@ -19,7 +28,20 @@ def self.details
...
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
+          FastlaneCore::ConfigItem.new(key: :version_properties_path,
+                                       description: 'Path to the version.properties file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:build_gradle_path, :project_name, :project_root_folder]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_alpha_version.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_alpha_version.rb [hunk 10]
```diff
@@ -19,7 +28,20 @@ def self.details
       end
 
       def self.available_options
-        # Define all options your action supports.
+        [
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_alpha_version.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_app_version.rb [hunk 11]
```diff
@@ -19,7 +28,20 @@ def self.details
...
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
+          FastlaneCore::ConfigItem.new(key: :version_properties_path,
+                                       description: 'Path to the version.properties file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:build_gradle_path, :project_name, :project_root_folder]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_app_version.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_release_version.rb [hunk 12]
```diff
@@ -19,7 +28,20 @@ def self.details
       end
 
       def self.available_options
-        # Define all options your action supports.
+        [
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_release_version.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_app_version.rb [hunk 13]
```diff
@@ -19,7 +28,20 @@ def self.details
       end
 
       def self.available_options
-        # Define all options your action supports.
+        [
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_app_version.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_alpha_version.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_get_alpha_version.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_finalize_prechecks.rb [hunk 14]
```diff
@@ -51,6 +59,18 @@ def self.available_options
...
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
+          FastlaneCore::ConfigItem.new(key: :version_properties_path,
+                                       description: 'Path to the version.properties file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:build_gradle_path, :project_name, :project_root_folder]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_finalize_prechecks.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_finalize_prechecks.rb [hunk 15]
```diff
@@ -51,6 +59,18 @@ def self.available_options
                                        description: 'Skips confirmation',
                                        type: Boolean,
                                        default_value: false), # the default value if the user didn't provide one
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_finalize_prechecks.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_current_branch_is_hotfix.rb [hunk 16]
```diff
@@ -24,7 +33,20 @@ def self.details
...
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
+          FastlaneCore::ConfigItem.new(key: :version_properties_path,
+                                       description: 'Path to the version.properties file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:build_gradle_path, :project_name, :project_root_folder]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_current_branch_is_hotfix.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_current_branch_is_hotfix.rb [hunk 17]
```diff
@@ -24,7 +33,20 @@ def self.details
       end
 
       def self.available_options
-        # Define all options your action supports.
+        [
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_current_branch_is_hotfix.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_completecodefreeze_prechecks.rb [hunk 18]
```diff
@@ -44,6 +52,18 @@ def self.available_options
...
+          FastlaneCore::ConfigItem.new(key: :build_gradle_path,
+                                       description: 'Path to the build.gradle file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:project_name, :project_root_folder, :version_properties_path]),
+          FastlaneCore::ConfigItem.new(key: :version_properties_path,
+                                       description: 'Path to the version.properties file',
+                                       type: String,
+                                       optional: true,
+                                       conflicting_options: [:build_gradle_path, :project_name, :project_root_folder]),
```
[the remark this answers is not in the document]
Comment (spencertransier) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_completecodefreeze_prechecks.rb]: Fixed in https://github.com/wordpress-mobile/release-toolkit/pull/519/commits/3e76dcf29247ce4ef912e15cd5ec3f8d94e91e1a

**Comment by spencertransier** [comment]
@jkmassel was AFK yesterday but messaged me that he completed testing this during the WCAndroid 15.9 finalization and it is good to merge.

**Pull request merged**